### PR TITLE
Add client methods

### PR DESCRIPTION
**Pull request by tarampampam**
| Q             | A
| ------------- | ---
| Bug fix?      |  No
| New feature?  | Yes 

## Description

- Package totally rewritten _(previous versions [available here](https://github.com/avto-dev/b2b-api-php))_

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote unit tests for my code
- [x] I have made changes in [CHANGELOG.md](https://github.com/avtocod/b2b-api-php/blob/master/CHANGELOG.md) file

### CODE REVIEW

File: src/Exceptions/BadRequestException.php [hunk 1]
```diff
@@ -26,20 +28,36 @@ class BadRequestException extends RuntimeException
...
-                                int $code = 0,
+                                ?string $message = null,
+                                ?int $code = null,
                                 ?Throwable $previous = null)
     {
         $this->http_request  = $http_request;
         $this->http_response = $http_response;
 
-        parent::__construct($message, $code, $previous);
+        $service_error_message = $http_response instanceof ResponseInterface
```
Comment (Reallife) [inline src/Exceptions/BadRequestException.php]: Maybe it makes sense to combine this with the code on line 52 (`$http_code = $http_response instanceof ResponseInterface ...`). No need to check instance of `$http_response` twice.

Something like this:
```php
$service_error_message = null;
$http_code = null;
if ($http_response instanceof ResponseInterface) {
    $service_error_message = $this->extractErrorMessageFromResponse($http_response);
    $http_code = $http_response->getStatusCode();
}
```

Comment (tarampampam) [reply src/Exceptions/BadRequestException.php]: Yay! Approved

File: src/ClientInterface.php [hunk 2]
```diff
@@ -52,4 +60,123 @@ public function devToken(string $username,
      * @return UserResponse
      */
     public function user(bool $detailed = false): UserResponse;
+
+    /**
+     * Retrieve balance information for report type.
+     *
+     * @param string $report_type_uid E.g.: `report_type@domain`
```
Comment (jetexe) [inline src/ClientInterface.php:67]: Unify 'e.g.' type format. `(e. g. %s)` perhaps

File: src/Responses/Entities/Balance.php [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+     */
+    public function getReportTypeUid(): string
+    {
+        return $this->report_type_uid;
+    }
+
+    /**
+     * Get balance entry type.
+     *
+     * E.g.: `DAY`, `MONTH`, `TOTAL`.
```
Comment (jetexe) [inline src/Responses/Entities/Balance.php]: Unify

File: src/Responses/Entities/ReportType.php [hunk 4]
```diff
@@ -430,6 +508,18 @@ public function getCleanOptions(): ?CleanOptions
         return $this->clean_options;
     }
 
+    /**
+     * Get report generation mode.
+     *
+     * E.g.: `TRANSACTIONAL`, `FAST_NON_TRANSACTIONAL`, `FAST_NON_BALANCE`.
```
Comment (jetexe) [inline src/Responses/Entities/ReportType.php]: Unify

File: src/Responses/UserReportsResponse.php [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    /**
+     * Get first report object.
+     *
+     * @return Report|null
+     */
+    public function first(): ?Report
+    {
+        if (isset($this->data[0])) {
```
Comment (jetexe) [inline src/Responses/UserReportsResponse.php]: `return $this->data[0]??null`?

File: src/Responses/UserResponse.php [hunk 6]
```diff
@@ -114,4 +117,38 @@ public function getData(): array
...
+     *
+     * @return User|null
+     */
+    public function getByUid(string $uid): ?User
+    {
+        $result = \array_values(\array_filter($this->data, function (User $user) use (&$uid): bool {
+            return $user->getUid() === $uid;
+        }));
+
+        return \count($result) === 1
```
Comment (jetexe) [inline src/Responses/UserResponse.php]: in phpdoc

> Get **first** user entry by UID.

But in realization you use
> `\count($result) === 1`

Should be `\count($result) > 0`

**Pull request merged**